Add PTZ support to Foscam camera component

skgsergio opened the pull request:
## Description
This PR extends the current `foscam` component to support PTZ through a service (like the `onvif` component).

I also took this opportunity to update the module for using the async API.

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io):** home-assistant/home-assistant.io#10625

### Demo
![Kapture 2019-10-06 at 13 31 29](https://user-images.githubusercontent.com/186349/66268443-ba6f0d80-e83d-11e9-96e7-583063e1efdc.gif)

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`.
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

skgsergio committed:
Add PTZ support to Foscam camera component

homeassistant commented:
Hi @skgsergio, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#27238).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare commented on homeassistant/components/camera/services.yaml:
```diff
@@ -93,3 +93,16 @@ onvif_ptz:
       description: "Zoom. Allowed values: ZOOM_IN, ZOOM_OUT"
       example: "ZOOM_IN"
+
+foscam_ptz:
```
Please move this to the foscam integration package services.yaml file.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -28,20 +60,70 @@
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up a Foscam IP Camera."""
-    add_entities([FoscamCam(config)])
+    _LOGGER.debug("Setting up the Foscam camera platform")
```
Please remove the logging. This is already logged by the core.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -28,20 +60,70 @@
         """Initialize a Foscam camera."""
+        super().__init__()
+
+        _LOGGER.debug("Importing dependencies")
```
Please remove.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -28,20 +60,70 @@
+    async def async_handle_ptz(service):
+        """Handle PTZ service call."""
+        movement = service.data.get(ATTR_MOVEMENT)
+        travel_time = service.data.get(ATTR_TRAVELTIME)
```
Please use `dict[key]` for required schema keys and keys with default schema values.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -28,20 +60,70 @@
+        _LOGGER.debug("Moving '%s' camera(s): %s", movement, entity_ids)
+
+        if not entity_ids:
+            target_cameras = all_cameras
```
Since we require entity_id in the service schema, it's unintuitive to target all cameras if no entity_id matches the camera entities.

We should just return here.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -28,20 +60,70 @@
+        DOMAIN, SERVICE_PTZ, async_handle_ptz, schema=SERVICE_PTZ_SCHEMA
+    )
+
+    _LOGGER.debug("Constructing the FoscamCam")
```
Please remove.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -28,20 +60,70 @@
     """An implementation of a Foscam IP camera."""
 
-    def __init__(self, device_info):
+    def __init__(self, hass, device_info):
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -28,20 +60,70 @@
+
+        _LOGGER.debug("Importing dependencies")
+
         from libpyfoscam import FoscamCamera
```
Please move 3rd party imports to the top of the module.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -28,20 +60,70 @@
         from libpyfoscam import FoscamCamera
 
-        super().__init__()
+        _LOGGER.debug("Setting up the Foscam camera component")
```
Please remove.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -50,6 +132,8 @@ def __init__(self, device_info):
         self._name = device_info.get(CONF_NAME)
         self._motion_status = False
 
+        _LOGGER.debug("Setting up the camera device @ '%s:%s'", ip_address, port)
```
We don't want side effects in init method. Please remove the logging.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -60,6 +144,18 @@ def __init__(self, device_info):
             if result == 0:
                 self._rtsp_port = response.get("rtspPort") or response.get("mediaPort")
 
+        _LOGGER.debug("RTSP port discovered: %s", self._rtsp_port)
```
As above. If we really need the logging message it should be done outside of init method.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -60,6 +144,18 @@ def __init__(self, device_info):
+        _LOGGER.debug("Camera '%s' added to hass", self._name)
+
+        if FOSCAM_DATA not in self.hass.data:
+            self.hass.data[FOSCAM_DATA] = {}
```
Please use `dict.setdefault` instead.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -111,6 +207,31 @@ def disable_motion_detection(self):
+            DIR_TOPRIGHT: self._foscam_session.ptz_move_top_right,
+            DIR_BOTTOMLEFT: self._foscam_session.ptz_move_bottom_left,
+            DIR_BOTTOMRIGHT: self._foscam_session.ptz_move_bottom_right,
+        }[movement]()
```
We're in async context here. The library doesn't support asyncio. We have to schedule sync I/O on the executor thread pool. Use `self.hass.async_add_executor_job`.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -111,6 +207,31 @@ def disable_motion_detection(self):
+        }[movement]()
+
+        if ret != 0:
+            _LOGGER.error("Error moving %s '%s': %s", movement, self._name, ret)
```
Should we return here?

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -111,6 +207,31 @@ def disable_motion_detection(self):
+        if ret != 0:
+            _LOGGER.error("Error moving %s '%s': %s", movement, self._name, ret)
+
+        sleep(travel_time)
```
We're not allowed to sync sleep in the event loop. Use `asyncio.sleep`.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -111,6 +207,31 @@ def disable_motion_detection(self):
+
+        sleep(travel_time)
+
+        ret, _ = self._foscam_session.ptz_stop_run()
```
See above.

MartinHjelmare reviewed (changes requested):
Please move all the side effects out of the entity init method.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -28,20 +60,70 @@
+            await camera.async_perform_ptz(movement, travel_time)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_PTZ, async_handle_ptz, schema=SERVICE_PTZ_SCHEMA
```
The service should be registered under the foscam integration domain, not the camera domain.

skgsergio replied:
On question here, if I register it under the foscam domain it doesn't filter in the developer utilities by the foscam cameras, it lists all entities without checking if it's a camera, that's why I did it (like the onvif component does).

MartinHjelmare replied:
That's a frontend issue. Custom services must be registered under the integration domain.

skgsergio replied:
Ok then, going to apply all changes. I fact this one made sense for me to do it under the foscam domain but when I faced that issue I ended doing it in the camera domain. Thanks.

skgsergio committed:
Address review comments:

 - Move service to foscam domain
 - Use `dict[key]` for required schema keys or with defaults
 - Fix sync operations in async context
 - Remove excessive logging

skgsergio commented:
Addresed all issues and re-tested. Not sure the procedure here, should I squash?

skgsergio committed:
Fix import order

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -17,6 +28,27 @@
 
 FOSCAM_COMM_ERROR = -8
 
+DOMAIN = "foscam"
```
Please don't define the `DOMAIN` constant in this module. That's ambiguous since this platform belongs to the camera domain and the integration domain is foscam. It's ok to import it from a const.py module as a different name eg `FOSCAM_DOMAIN`.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -60,6 +133,12 @@ def __init__(self, device_info):
             if result == 0:
                 self._rtsp_port = response.get("rtspPort") or response.get("mediaPort")
```
Can we move this and the request out of init? It's not allowed to do sync I/O in async context.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -60,6 +133,12 @@ def __init__(self, device_info):
+    async def async_added_to_hass(self):
+        """Handle entity addition to hass."""
+        self.hass.data.setdefault(FOSCAM_DATA, {})
+        self.hass.data[FOSCAM_DATA].setdefault(ENTITIES, [])
```
```py
entities = self.hass.data.setdefault(
    FOSCAM_DATA, {}).setdefault(ENTITIES, [])
entities.append(self)
```

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -111,6 +190,37 @@ def disable_motion_detection(self):
+                DIR_TOPRIGHT: self._foscam_session.ptz_move_top_right,
+                DIR_BOTTOMLEFT: self._foscam_session.ptz_move_bottom_left,
+                DIR_BOTTOMRIGHT: self._foscam_session.ptz_move_bottom_right,
+            }[movement]
```
Please separate the dict lookup from the call on different lines to make it more readable.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -1,11 +1,22 @@
 """This component provides basic support for Foscam IP cameras."""
 import logging
+from asyncio import sleep
```
Please import `asyncio`. That makes it clear what sleep we're using below in the code.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -28,26 +60,67 @@
+        self._name = device_info[CONF_NAME]
         self._motion_status = False
 
         self._foscam_session = FoscamCamera(
```
Create the `FoscamCamera` instance in `async_setup_platform` and pass it to the entity after checking the rtsp port.

skgsergio committed:
Move all the initialization to setup_platform and fix motion detection status logic

skgsergio commented:
Moved al initialization to `async_setup_platform` now `__init__` doesn't contain logic.
Moved constants to `const`.
I saw the motion detection logic wasn't ok so I fixed it too in this commit.
Addressed the rest of the comments.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -97,19 +196,56 @@ def enable_motion_detection(self):
+        _LOGGER.debug("PTZ action '%s' on %s", movement, self._name)
+
+        movement_function = {
+            DIR_UP: self._foscam_session.ptz_move_up,
```
We could break out the dict to a module level constant if we write the method names as strings in the values of the dict. Then use `getattr` to get the method.

```py
MOVEMENT_ATTRS = {
    DIR_UP: "ptz_move_up",
    ...
}

...

movement_attr = MOVEMENT_ATTRS[movement]
movement_function = getattr(self._foscam session, movement_attr)
```

skgsergio replied:
I thought about doing this but didn't know if `getattr` was acceptable, I've seen some rejection previously, but yes this is cleaner so let's do it.

skgsergio committed:
Move function dictionary out of the function.

MartinHjelmare commented on homeassistant/components/foscam/camera.py:
```diff
@@ -15,7 +30,32 @@
+
+DEFAULT_TRAVELTIME = 0.125
+
+DIR_UP = "UP"
```
Should we make these strings lowercase snakecase? That's our standard for other strings that the user needs to handle.

skgsergio replied:
Done.

skgsergio committed:
Change user input to lowercase snake case

skgsergio committed:
Change user input to lowercase snake case

MartinHjelmare commented on homeassistant/components/foscam/services.yaml:
```diff
@@ -0,0 +1,12 @@
+      example: 'camera.living_room_camera'
+    movement:
+      description: "Direction of the movement. Allowed values: up, down, left, right, top_left, top_right, bottom_left, bottom_right."
+      example: 'UP'
```
"up"

skgsergio replied:
🤦‍♂

skgsergio committed:
Fix service example value

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Can be merged when build passes.

frenck reviewed (changes requested):
Please exclude `homeassistant/components/foscam/const.py` from our test coverage by adding it to `.coveragerc`.

skgsergio committed:
Omit foscam const module from code coverage tests

frenck reviewed (approved):
Thanks!

frenck commented:
Thank you for your contribution thus far! 🎖 One final thing: Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

skgsergio committed:
Add myself to foscam codeowners

skgsergio commented:
Done 👍 and thank everyone for the project and the awesome help with the PR (next time I'll have the guidelines more clear 😂)

frenck commented:
Well, the opposite is the case: Thank you sir! 🎉
